bookmarks/services/parser.py

```diff
@@ -12,18 +14,81 @@ class NetscapeBookmark:
     tag_string: str
 
 
+class BookmarkParser(HTMLParser):
+
+    tag_stack: List[Optional[str]] = []
+    tag: Optional[str] = None
+    bookmarks: List[NetscapeBookmark] = []
+    bookmark: Optional[NetscapeBookmark] = None
+    folder: Optional[str] = ''
+    description: Optional[str] = None
+
+    def handle_starttag(self, tag: str, attrs: list):
+        name = 'handle_start_' + tag.lower()
+        if name in dir(self):
+            getattr(self, name)({k.lower(): v for k, v in attrs})
+        self.tag = tag
+
+    def handle_endtag(self, tag: str):
+        name = 'handle_end_' + tag.lower()
+        if name in dir(self):
+            getattr(self, name)()
+        self.tag = None
+
+    def handle_data(self, data):
+        name = f'handle_{self.tag}_data'
+        if name in dir(self):
+            getattr(self, name)(data)
+
+    def handle_start_dl(self, attrs: Dict[str, str]):
+        print('<DL>')
+        self.tag_stack.append(None)
```

Comment:
This part, and the related logic in `handle_h3_data` and `handle_a_data` seems a bit convoluted. I think it might be easier to just add a tag to the stack when encountering an `<h3>`, and then to remove a tag from the stack when encountering and `</dl>` - unless the stack is already empty.